File: src/main/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/applikasjon/modell/bransjeprogram/BransjeEllerNæring.java
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.modell.bransjeprogram;

import io.vavr.control.Either;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.modell.Næring;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.Statistikkategori;

public class BransjeEllerNæring {
  public final Either<Bransje, Næring> verdi;

  public BransjeEllerNæring(Bransje bransje) {
    this.verdi = Either.left(bransje);
  }

  public BransjeEllerNæring(Næring næring) {
    this.verdi = Either.right(næring);
  }

  public Statistikkategori getStatistikkategori() {
    if (verdi.isLeft()) {
      return Statistikkategori.BRANSJE;
    } else {
      return Statistikkategori.NÆRING;
    }
  }

  public boolean isBransje() {
    return verdi.isLeft();
  }

  public Bransje getBransje() {
    return verdi.getLeft();
  }

  public String navn() {
    return this.isBransje() ? verdi.getLeft().getNavn() : verdi.get().getNavn();
  }

  public Næring getNæring() {
    return verdi.get();
  }
}

File: src/main/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/applikasjon/domenemodeller/bransjeprogram/BransjeEllerNæring.kt
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domenemodeller.bransjeprogram

import io.vavr.control.Either
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domenemodeller.BedreNæring
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domenemodeller.Statistikkategori

class BransjeEllerNæring {
    val verdi: Either<Bransje, BedreNæring>

    constructor(bransje: Bransje) {
        verdi = Either.left(bransje)
    }

    constructor(næring: BedreNæring) {
        verdi = Either.right(næring)
    }

    val statistikkategori: Statistikkategori
        get() = if (verdi.isLeft) {
            Statistikkategori.BRANSJE
        } else {
            Statistikkategori.NÆRING
        }
    val isBransje: Boolean
        get() = verdi.isLeft

    fun getBransje(): Bransje {
        return verdi.left
    }

    fun navn(): String {
        return if (isBransje) verdi.left.navn else verdi.get().navn
    }

    val næring: BedreNæring
        get() = verdi.get()
}
